src/vopr/main.go

```diff
@@ -0,0 +1,782 @@
+func (output *vopr_output) hash_stack_trace(message *vopr_message) {
+	// Check if there is a stack trace
+	if len(output.stack_trace) > 0 {
+		check_sum := sha256.Sum256(output.stack_trace)
+		output.stack_trace_hash = fmt.Sprintf("%x", check_sum)
+		log_debug("The stack trace has been hashed: "+output.stack_trace_hash, message.hash[:])
+	} else {
+		output.stack_trace_hash = ""
+		log_debug("No stack trace was found to hash", message.hash[:])
+	}
+}
+
+// The limited buffer ensures that only a specified number of bytes can be written to the buffer.
+type size_limited_buffer struct {
+	byte_budget int
+	// The buffer being wrapped
+	buffer *bytes.Buffer
+	// This channel is used to keep track of when the VOPR reaches its maximum allowed output
+	size_reached chan bool
+}
+
+// The limited buffer honours the Write function in such a way that only a specified number of
+// bytes can be written to the buffer. If the maximum number of bytes has been written then no
+// more will be added and a value of true is passed to the size_reached channel which acts as a
+// signal to kill the process that is doing the writing.
+func (limited_buffer *size_limited_buffer) Write(byte_array []byte) (int, error) {
+	space_left := limited_buffer.byte_budget
+	limited_buffer.byte_budget -= len(byte_array)
+	if limited_buffer.byte_budget <= 0 {
+		limited_buffer.size_reached <- true
+		if space_left > 0 {
+			return limited_buffer.buffer.Write(byte_array[0 : space_left-1])
+		} else {
+			return 0, nil
+		}
+	} else if space_left > 0 {
+		return limited_buffer.buffer.Write(byte_array)
+	} else {
+		return 0, nil
+	}
+}
+
+// Reads incoming messages (ensuring that only the correct number of bytes are read), decodes and
+// validates them before adding the messages to the vopr_message_channel.
+func handle_connection(track_connections chan bool, connection net.Conn, vopr_message_channel chan vopr_message) {
+	// When this function completes, close the connection and decrement the connections count.
+	defer func() {
+		connection.Close()
+		<-track_connections
+		log_debug("Connection closed", nil)
+	}()
+
+	var input vopr_message_byte_array
+
+	// Only reply if everything is as expected.
+	bytes_read, error := connection.Read(input[:])
+	if error != nil {
+		error_message := fmt.Sprintf("Client closed unexpectedly: %s", error.Error())
+		log_error(error_message, nil)
+		return
+	}
+	if bytes_read != LENGTH_OF_VOPR_MESSAGE {
+		log_error("The input was longer or shorter than expected", nil)
+		return
+	}
+
+	// Decodes the byte array into a vopr_message
+	message, decoding_error := decode_message(input)
+
+	if decoding_error == nil {
+		log_message := fmt.Sprintf(
+			"bug: %d commit: %x seed: %d",
+			message.bug,
+			message.commit,
+			message.seed,
+		)
+		log_info(log_message, message.hash[:])
+
+		// Checks if there is capacity to process the message
+		if len(vopr_message_channel) < MAX_QUEUING_MESSAGES {
```

Comment:
This seems like a TOCTOU (time of check vs time of use) error point:
* g1: `len(vopr_message_channel) < MAX_QUEUING_MESSGES` passes
* g1: is preempted (e.g. trying to `connection.Write` to a slow client, or from the scheduler)
* g2-N: (other connections) *fill up the message channel*
* g1: is scheduled and does `go write_to_vopr_message_channel()` when the channel is full now.

This could end up with `> MAX_QUEUING_MESSAGES` connections getting sent the "1" below and having their message queued up to be sent by the `go write_to_vopr_message_channel()`. 

While this doesn't block the connection (nice throughput optimization), it allows an attacker to queue up / spawn an unbounded amount of goroutines and messages between the len() check and the go statement. This short time frame could be expanded by having the clients reader be slow, making the `connection.Write()` possibly suspend and take longer before sending on the channel to bump the len.

One possible fix is to do a non-blocking send on the `vopr_message_channel`, *then* write "1". This would be fast (not block the connection) and could detect the full channel immediately (the len check + push operations are atomic) which removes the window for side-stepping backpressure.

```go
select {
case vopr_message_channel <- message:
    _, error = conneciton.Write([]byte("1"))
    ...
default:
    log_info("Too many messages already queued, dropping ...")
    ...
}
```